pkg/ccl/streamingccl/logical/lww_row_processor_test.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// BenchmarkLastWriteWinsInsert is a microbenchmark that targets overhead of the
+// SQL layer when processing the INSERT query of the LWW handler. It mocks out
+// the KV layer by injecting an error during KV request evaluation.
+func BenchmarkLastWriteWinsInsert(b *testing.B) {
+	defer leaktest.AfterTest(b)()
+	defer log.Scope(b).Close(b)
+
+	var discardWrite atomic.Bool
+	// key will be set later, before flipping discardWrite to true.
+	var key roachpb.Key
+	var numInjectedErrors atomic.Int64
+	injectedErr := kvpb.NewError(errors.New("injected error"))
+
+	ctx := context.Background()
+	srv, sqlDB, kvDB := serverutils.StartServer(b, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+		Knobs: base.TestingKnobs{
+			Store: &kvserver.StoreTestingKnobs{
+				TestingRequestFilter: func(_ context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
+					// In order to focus the benchmark on the SQL overhead of
+					// evaluating INSERT stmts via the internal executor, we'll
+					// inject errors for them so that the processing of the KV
+					// request stops once it reaches the KV client.
+					if !discardWrite.Load() || !ba.IsWrite() || len(ba.Requests) != 1 {
+						return nil
+					}
+					req, ok := ba.GetArg(kvpb.Put)
+					if !ok {
+						return nil
+					}
+					if !req.(*kvpb.PutRequest).Key.Equal(key) {
+						return nil
+					}
+					numInjectedErrors.Add(1)
+					return injectedErr
+				},
+			},
+		},
+	})
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
+
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+	tableName := "tab"
+	runner.Exec(b, "CREATE TABLE tab (pk int primary key, payload string)")
+	runner.Exec(b, "ALTER TABLE tab ADD COLUMN crdb_internal_origin_timestamp DECIMAL NOT VISIBLE DEFAULT NULL ON UPDATE NULL")
+
+	desc := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), "defaultdb", tableName)
+	tableDescs := make(map[string]descpb.TableDescriptor, 1)
+	tableDescs["defaultdb.public."+tableName] = *desc.TableDesc()
+	rp, err := makeSQLLastWriteWinsHandler(ctx, s.Codec(), s.ClusterSettings(), tableDescs)
+	require.NoError(b, err)
+
+	// Insert a single row into the table which we'll attempt to overwrite via
+	// the row processor.
+	// TODO(yuzefovich): I don't quite understand why this is needed - it
+	// appears as if the BatchRequest produced for the rp's INSERT query changes
+	// without it.
```

Comment:
I believe this is because a ConditionalPut is generated when there is no existing row. So if you don't include this, your first transaction in the loop below will be a ConditionalPut rather than a Put.  On the first run `b.N` will be `1` so you'll only ever get the ConditionalPut and your request filter will never inject the error.